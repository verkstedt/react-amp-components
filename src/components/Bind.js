import React, { Component, Fragment } from 'react'
import { oneOfType, arrayOf, node } from 'prop-types'
import { Helmet } from 'react-helmet'
import getBindAttributeName from '../utils/getBindAttributeName'

class Bind extends Component {
  static propTypes = {
    children: oneOfType([
      arrayOf(node),
      node
    ])
  }

  static defaultProps = {
    children: []
  }

  render() {
    const { children, ...props } = this.props
    const bindAttributes = Object.keys(props).reduce((acc, prop) => ({
      ...acc,
      [getBindAttributeName(prop)]: props[prop]
    }), {})

    return (
      <Fragment>
        <Helmet key="helmet">
          <script async custom-element="amp-bind" src="https://cdn.ampproject.org/v0/amp-bind-0.1.js" />
        </Helmet>
        {React.Children.map(children, child => (
          <child.type {...{ ...child.props, ...bindAttributes }} >
            {child.props.children}
          </child.type>
        ))}
      </Fragment>
    )
  }
}

export default Bind
