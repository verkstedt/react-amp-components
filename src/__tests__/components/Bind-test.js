import React from 'react'
import Bind from '../../components/Bind'
import Helmet from '../../utils/Helmet'
import { renderComponent } from '../test-utils'

describe('Bind', () => {
  it('works', () => {
    const res = renderComponent(
      <Bind text="'bind ' + foo">
        <div>test</div>
      </Bind>
    )
    expect(res.toJSON()).toMatchSnapshot()
  })

  it('works with multiple children', () => {
    const res = renderComponent(
      <Bind text="'bind ' + foo">
        <div>test</div>
        <p>another test</p>
      </Bind>
    )
    expect(res.toJSON()).toMatchSnapshot()
  })

  it('works with props with brackets', () => {
    const props = { '[text]': 'foo' }
    const res = renderComponent(
      <Bind {...props} >
        <p>test</p>
      </Bind>
    )
    expect(res.toJSON()).toMatchSnapshot()
  })

  it('works with child with children', () => {
    const res = renderComponent(
      <Bind className="'bind ' + foo">
        <div id="id" className="class">
          <p>One</p>
          <p>Another</p>
        </div>
      </Bind>
    )
    expect(res.toJSON()).toMatchSnapshot()
  })

  it('injects the right script tag', () => {
    renderComponent(
      <Bind>
        <p>Just test the script.</p>
      </Bind>
    )
    Helmet.canUseDOM = false
    const staticHead = Helmet.renderStatic()
    expect(staticHead.scriptTags).toMatchSnapshot()
  })

  it('passes down the right props', () => {
    const props = {
      '[text]': "'test ' + foo"
    }
    const bind = renderComponent(
      <Bind text="'test ' + foo">
        <p>bind me.</p>
      </Bind>
    )
    expect(bind.toJSON().props).toMatchObject(props)
  })
})

