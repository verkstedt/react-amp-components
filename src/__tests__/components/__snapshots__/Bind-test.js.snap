// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bind injects the right script tag 1`] = `
Array [
  Object {
    "async": "",
    "custom-element": "amp-bind",
    "src": "https://cdn.ampproject.org/v0/amp-bind-0.1.js",
  },
]
`;

exports[`Bind works 1`] = `
<div
  [text]="'bind ' + foo"
>
  test
</div>
`;

exports[`Bind works with child with children 1`] = `
<div
  [class]="'bind ' + foo"
  className="class"
  id="id"
>
  <p>
    One
  </p>
  <p>
    Another
  </p>
</div>
`;

exports[`Bind works with multiple children 1`] = `
Array [
  <div
    [text]="'bind ' + foo"
  >
    test
  </div>,
  <p
    [text]="'bind ' + foo"
  >
    another test
  </p>,
]
`;

exports[`Bind works with props with brackets 1`] = `
<p
  [text]="foo"
>
  test
</p>
`;
